### ISSUE TITLE
AWS Fargate GPU Support: When is GPU support coming to fargate?

### ISSUE THREAD

**Issue by mbnr85** [aws/containers-roadmap#88]
**Tell us about your request**
What do you want us to build?

**Which service(s) is this request for?**
This could be Fargate, ECS, EKS, ECR

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
What outcome are you trying to achieve, ultimately, and why is it hard/impossible to do right now? What is the impact of not having this problem solved? The more details you can provide, the better we'll be able to understand and solve the problem.

**Are you currently working around this issue?**
How are you currently solving this problem? 

**Additional context**
Anything else we should know?

**Attachments**
If you think you might have additional information that you'd like to include via an attachment, please do - we'll take a look. (Remember to remove any personally-identifiable information.)

**Comment by pauncejones** [comment]
Hi There, can you give us more details about your use case? Instance type, CUDA version, and more info about what you're trying to do - workload, etc.? Thanks.

**Comment by mbnr85** [comment]
We would like to run object detection on Fargate.

Setup: 
CUDA version 9.0, 9.1 (both work)
Instance type p2.xlarge
Algorithm: Object detection
Input: Frame
Output: Metadata preferably json with coordinates and confidence.
TPS: 10 frames/sec

Does Fargate have some concept of reserved instance discounts in EC2 or Sustained usage discounts?

**Comment by FernandoMiguel** [comment]
> Does Fargate have some concept of reserved instance discounts in EC2 or Sustained usage discounts?

No

**Comment by mikaelhg** [comment]
I have a similar use case. I'd like to run deep learning inference tasks on CUDA-capable GPUs on Fargate, and pay per second of usage. 

The specific use case is inference tasks which are run fairly seldom, but need to respond in seconds, rather than minutes. In other words, waiting a few minutes for an EC2 instance to boot up, just doesn't cut the mustard. But neither does the application need to be taking up a GPU 24/7 unproductively, just to run the inference job for a minute or two, twice a day.

**Comment by juve** [comment]
I also have an inference use-case where we would like to be able to autoscale inference sqs workers in Fargate. We originally tried to use ECS, but found it too cumbersome to scale both the containers and the EC2 instances, so we are currently just using EC2 instances with an autoscaling group. We considered using Sagemaker, but that will require some engineering effort for us to adapt our architecture and models.

**Comment by aysark** [comment]
I'd be interested in this too and have similar usecases as above.

**Comment by gfodor** [comment]
I have a use case for this too, where we want to spin up GPU resources to do live video streaming but relinquish those completely after the stream ends, with minimal start up time.

**Comment by prameshbajra** [comment]
@mbnr85 I too am trying to do object detection on fargate. 
Is this even possible (for now)? Have you found anything? What did you do in your case?

**Comment by ngander-amfam** [comment]
When training data science models our workloads can take advantage of GPU compute. To start those workloads will run in ECS although eventually we’d likely migrate those to EKS. We’d like to be able to use Fargate to run GPU accelerated workloads but that is not currently supported. Does AWS have GPU compute on the Fargate roadmap, and if so, is there any timeline that can be shared?

**Comment by tomfranken** [comment]
Also interested for machine learning...

**Comment by romanovzky** [comment]
Interested for ML training and inference as well. The overhead to transfer to sagemaker is too high, we just train models on EC2 GPU boxes and then use CPU runtime for inference on Fargate instances. However, some models would benefit from GPU at inference time (namely those trained on CUDA specific implementations, which as of now we are not using for lack of inference infrastructure). The inference use case is sporadic, such that a full-time EC2 box is too pricey.

**Comment by prameshbajra** [comment]
@romanovzky We both are on the same boat I guess. I too am in a similar situation.

**Comment by ashirgao** [comment]
I too am looking forward for this feature.

**My use-case:**

I need to run jobs that benefit from GPU acceleration (mostly model inference and some CPU bound tasks eg. embedding clustering, DB insertions etc.). Each job takes around 10-15 mins on a p2.xlarge. I receive 100-120 such jobs through the day (get 8-10 jobs in the span of 30 sec at max).  

**My requirement:**

A server-less GPU container solution.

**My current solution:**

My GPU utilizing containers run as custom Sagemaker training jobs. 

Advantages:
+ With my increased Sagemaker limit on p2.xlarge systems, I can have 20 jobs running in parallel. And 0 idle cost. So, sort of server-less GPU containers :)
+ Per-second billing.
+ My containers have minimal Sagemaker specific code and hence can be easily run on EC2, ECS or even my own desktop system.

Disadvantages:
- Sagemaker actually spawns a new instance for my container. This results in longer wait times. (Usually 2x Fargate wait times.)
- Need to add additional logic in my lambda function that triggers Fargate jobs and Sagemaker jobs separately.

**Comment by ctmckee** [comment]
Also.... 
Some machine learning models require GPU support for predictions (they will not predict on CPU). 

For example (an InternalError that can occur when attempting to get a RefineNet predictions on CPU):
InternalError: The CPU implementation of FusedBatchNorm only supports NHWC tensor format for now.

I too support GPU support with Fargate

**Comment by Zirkonium88** [comment]
We would like to call from a Docker container (RStudio) several others for a distributed deep/machine learning training using Fargate/AWS Batch. The results should saved on S3 and wrote back to RStudio Docker container. So far Fargate shows no support for GPUs.

**Comment by richarms** [comment]
I would also like to launch GPU containers from Fargate. I have two use-cases: 1. spawning powerful deep learning Jupyterhub development environments for our machine-learning group's researchers that will effortlessly disappear when the individual Jupyterhub kernel is killed. 2. Infrequent, quickly-scaled, deep (i.e. the use of GPU is justified) inference tasks. 

a thought: for 2., I hadn't thought of using the suggestion above of an auto-scaling EC2 group (that presumably then use something like a scripted docker-machine command to provision the instance, and launch a kernel container) to run the GPU containers, but this seems like a nasty, expensive (in time and currency) hack for what should be a bit more elegant.

**Comment by ClaasBrueggemann** [comment]
Any news on this?

**Comment by prameshbajra** [comment]
@ClaasBrueggemann I dont think they will provide this anytime soon. AWS is heavily promoting SageMaker now and many cases that's the way to go. :)

**Comment by jl-DaDar** [comment]
what about for 3d model rendering? we aren't needing this for machine learning.

**Comment by gautamkmr** [comment]
+1 for this support.

**Comment by prameshbajra** [comment]
> what about for 3d model rendering? we aren't needing this for machine learning.

In that case getting a GPU instance like P2, G3 etc might help?
Amazon won't be providing GPUs in fargate I believe.

**Comment by srinivaspype** [comment]
Any SLA for this? Currently Fargate implementation provides general-purpose CPU cycle speed 2.2GHz- 2.3GHz for us and not  capable of running CPU/GPU critical applications.

**Comment by Adiii717** [comment]
Fargate does not support GPU and we can expect nearly in future.

> In Closing
> Fargate helped us solve a lot of problems related to real-time processing, including the reduction of operational overhead, for this dynamic environment. We expect it to continue to grow and mature as a service. Some features we would like to see in the near future include GPU support for our GPU-based AI Engines and the ability to cache container images that are larger for quicker “warm” launch times.
https://aws.amazon.com/blogs/architecture/building-real-time-ai-with-aws-fargate/

**Comment by depthwise** [comment]
FWIW, it'd be great to run a typical deep learning experiment queue on something like this. Upload code+configs to S3. Lambda picks up, stuffs it into a container, training runs to completion and saves back to S3. Super simple, very scalable.

**Comment by prameshbajra** [comment]
> FWIW, it'd be great to run a typical deep learning experiment queue on something like this. Upload code+configs to S3. Lambda picks up, stuffs it into a container, training runs to completion and saves back to S3. Super simple, very scalable.

Sounds much more like something that **sagemaker** would do.

**Comment by mrichman** [comment]
What is the status of this? I'm very interested in CUDA support in Fargate tasks.

**Comment by ndtreviv** [comment]
I want to use GPU-optimised faiss training algorithms on fargate. I'm not training or running a model, I'm just training an HNSW index on faiss.

**Comment by siobhansabino** [comment]
I have a slightly different use case in that it doesn't involve AI/ML at all. I need to provide my data science team with GPUs in a serverless context for massive calculations that run better on GPUs than CPUs. They run ad hoc containers in an ad hoc manner, so Fargate makes the most sense in enabling them to ship their containers and perform whatever they need instead of needing to max out their local machine. No other AWS service meets this need without requiring extra operational help which is what we are trying to avoid to allow the team to retain ownership over their work.

**Comment by pagameba** [comment]
We would like to be able to use an on-demand GPU with headless Chromium for scheduling jobs to render WebGL image filters implemented as shaders.  Currently we are using the SwiftShader in a lambda function for this because we only need to do this a few times a day but need lower latency than an EC2 auto-scaling group. SwiftShader is very slow, however, and is not identical to running on an actual GPU, causing some image quality issues.  Having GPU support in Fargate would allow us to spin up ondemand containers to service rendering jobs with overall higher performance than the current lambda solution, while keeping operational costs aligned with actual usage.

Elastic GPU support in lambdas would be amazing too :)

**Comment by edgBR** [comment]
We have a similar use case to @Zirkonium88 

We have a p3.8 large instance where we have rstudio teams and we would like to downsize the instance quite a lot to use the kubernetes launcher feature of RStudio. We are using EKS  backed with Fargate to launch our jupyterlab sessions and rstudio sessions but some of our users will need GPU acceleration for prototyping.

**Comment by craiglytle** [comment]
+1 for GPU on Fargate.  ML application.
(Is it rude to point out that Azure offers this in their container instance service in preview now?)

**Comment by fnandocontreras** [comment]
+1 GPU support on Fargate:
Usecase: I want create a task definition using Fargate for training automation of a DeepLearning model in tensorflow

**Comment by shlomi-viz** [comment]
I use AWS Batch for that

**Comment by craiglytle** [comment]
@shlomi-viz 
Does AWS Batch work well for that use case?  
Can you easily launch containers in Batch?  
How fast do the new EC2 instances launch; seconds or minutes?

**Comment by lefnire** [comment]
@craiglytle for my part, it's minutes: anywhere between 2 & 30. Extremely variable, with a huge high-end; a main reason I'm interested in this ticket (though I still need to try raw ECS). I do have a Queue with Spot Instances tried first, On-demand instances fallback; so that could contribute to the launch time.

**Comment by shlomi-viz** [comment]
I tried to implement this: https://aws.amazon.com/blogs/aws/aws-ecs-cluster-auto-scaling-is-now-generally-available/
but it require you to "play" a bit to find your best values for scale in and out, for me that was too slow. 
AWS Batch can scale much faster and smarter, BUT it evaluate the scaling every 10 minutes, meaning that every time it will take at least 10 minutes until you start to scale up. From my experience once it started to scale it will be quite fast if you have many tasks in the queue.
You well need to create and mange more resources then running with Fargate, but this is a one time cost.
From me launching a container was basically the same,  call `submit_job` instead of `run_task`.
I have tested it with ~1K jobs/task and it worked fine. 

Happy to add more info, if you need @craiglytle

**Comment by jpujari-ctx** [comment]
+1 GPU support on Fargate

Usecase: need to run data analysis using rapids library

Current Solution:
Use Dask and ECS. I have a ECS GPU service for spinning up ECS GPU workers using ECS capacity providers. 

Drawbacks:
Each task spins up a GPU EC2 instance, which takes 15-20 minutes.

**Comment by dslove** [comment]
+1 00000000...
Sagemaker doesn't fit into my case because I need to do a lot of work for conversion. EC2 is just to cumbersome to use - I have to manage a lot of stuff, like autoscaling.

**Comment by andreasunterhuber** [comment]
+1 👍

**Comment by genifycom** [comment]
We are using CUDA for LIDAR processing. Latest CUDA version would be fine.

The pipeline needs to process thousands of files, loading up the GPU memory and performing analysis and transformation of said data.

We do need a LOT of GPU memory for these data sets, but really any move on the Fargate/GPU support would help at this point.

**Comment by dheerajmpai** [comment]
Same here, this seems to be a long awaited demand. Looking for Fargate or Lambda support for GPU or Inf Instances.

**Comment by masoom** [comment]
Much awaited feature. Tagging @nathanpeck for some attention :)

**Comment by dheerajmpai** [comment]
Is there any specific technical reason why Fargate does not support GPU Instances?

**Comment by prameshbajra** [comment]
> Is there any specific technical reason why Fargate does not support GPU Instances?

@dheerajmpai My company asked this to AWS Support team. They mentioned that they were working on it. However, we were not provided with any dates. Hope they bring this soon.

**Comment by shkr** [comment]
+1

**Comment by jonrau1** [comment]
+1 on this - running a seperate EC2-based ECS Cluster when every single other thing is on Fargate is needless overhead, we want to  retrain some IP Insights and DGL models on push from a Fargate-based API server.

**Comment by NIMO-Industries** [comment]
+1

**Comment by icalvete** [comment]
+1

**Comment by TimDommett** [comment]
Potential use case for us is "pixel streaming" our educational application for students who don't have the required hardware, where the GPU resources will need to scale up and down unexpectedly as students join and leave the online application.

**Comment by sinwar** [comment]
Please add GPU support for Fargate, we need it for real-time prediction which runs way faster on GPU. Without GPU customer will face high latency issues.

**Comment by prameshbajra** [comment]
> Please add GPU support for Fargate, we need it for real-time prediction which runs way faster on GPU. Without GPU customers will face high latency issues.

We had a similar use case too. We are using SageMaker now and it is literally amazing. The `Batch Transform` feature is literally amazing. We are running 1000+ predictions under 500 seconds.
I would suggest going with SageMaker as waiting for this feature to come might take a long long time.

**Comment by genifycom** [comment]
Thanks for the suggestion.

**Comment by TimDommett** [comment]
> > Please add GPU support for Fargate, we need it for real-time prediction which runs way faster on GPU. Without GPU customers will face high latency issues.
> 
> We had a similar use case too. We are using SageMaker now and it is literally amazing. The `Batch Transform` feature is literally amazing. We are running 1000+ predictions under 500 seconds.
> I would suggest going with SageMaker as waiting for this feature to come might take a long long time.

Thanks for the suggestion @prameshbajra we will definitely give this a try.

**Comment by pkiv** [comment]
+1, would appreciate GPU support on fargate. Even if it was fractional GPU support. (my fargate tasks only need 10-20% of a T4).

**Comment by PhilCliv** [comment]
Need GPU support as well

**Comment by lefnire** [comment]
Y'all, it's not gonna come - don't hold your breath. There's lots of GPU-based options outside Fargate: SageMaker [Batch Transform](https://docs.aws.amazon.com/sagemaker/latest/dg/batch-transform.html) or [Transform Job](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/sagemaker/create-transform-job.html). [AWS Batch](https://aws.amazon.com/batch/) - maybe fractional GPU support via containers sharing an instance / cluster? Mix/matching instance with GPU via [Elastic GPUs](https://aws.amazon.com/ec2/elastic-graphics/) / [Elastic Inference](https://aws.amazon.com/machine-learning/elastic-inference/). Efficient GPUs eg [Trainium](https://aws.amazon.com/machine-learning/trainium/) or [Inferentia](https://aws.amazon.com/machine-learning/inferentia/). SageMaker can share GPU usage via multi-model endpoints.

So - EC2, (ECS?), Batch, SageMaker (Batch Transform, Transform Jobs), and mix/matching GPU on that list via Elastic [Inference|GPU], Trainium, Inferentia, etc. All that's to say, Fargate GPU support might not come - because it's pretty much not necessary (covered by other services).

**Comment by whillas** [comment]
@lefnire none of other services scale to zero after job is run. Thats the main point of Fargate. Can do this in GKE, setup node group with GPU nodes that will scale to zero once job is finished. Maybe we'll just run this workload over there...

**Comment by lefnire** [comment]
@whillas that's an important feature to me too, so correct me if I'm wrong in my explanation of how they _do_ scale to 0. Batch & SageMaker Batch Transform are fire-and-forget. They're not auto-scaling based on load, instead you manually submit a job - new container - run till completion - die. As opposed to SageMaker Endpoints which is a managed auto-scaling environment, which indeed can't scale to 0 (which I hate - honestly I think we should be requesting scale-to-0 SageMaker endpoints rather than Fargate GPUs).

The bummer here is cold-start loading of resources. Eg, loading a huggingface/transformers model into memory before running inference can take quite some time. I'm not sure if there's an equivalence in Batch / SM:BT to the way freshly-completed Lambda functions can keep variables loaded outside `handler()`?

So my thought is: rather than thinking of an auto-scaling Fargate cluster, just submit manual one-offs via Batch / SM:BT. I actually didn't know that Fargate scales to 0... that does alter my opinion here. I am crossing my fingers they listen, just not holding breath they do.

**Comment by whillas** [comment]
@lefnire yeah you are right about the Batch Transform, it does [scale up and down for standalone  jobs](https://www.youtube.com/watch?v=Z9FtrRq0rc0&ab_channel=AmazonWebServices) and it suggested that you can trigger these jobs via lambda functions, which is another requirement that most people have i.e. **orchestration** in production which is why i was looking at this in the first place.

My gripes with it are :
- that forces you to stick your data in S3 (our is in a DB but i guess we can wirte a sperate ETL to dump to a bucket). 
- it does make some assumption abotu how your going to process the data i.e. record by record (we need to have access to all of the records/results at once... which i think SM:BT can do but nto sure about AWS Batch?)
- Sagemaker instance cost more :(

Thanks  for being a sounding board! 
(Its a pitty AWS's forums are useless)

**Comment by lefnire** [comment]
Just saw [Serverless NLP Inference on Amazon SageMaker with Transformer Models from Hugging Face](https://towardsdatascience.com/serverless-nlp-inference-on-amazon-sagemaker-with-transformer-models-from-hugging-face-4843609a7451). Seems SageMaker now has [Serverless Inference](https://docs.aws.amazon.com/sagemaker/latest/dg/serverless-endpoints.html). For me at least, this solves the issue for which I'd subscribed to this thread. I'll report back after fiddling with it, if anyone cares

**Comment by heukirne** [comment]
@lefnire , ECS still had the advantage of using Spot Instance price for running inference for batch purposes. Sagemaker Batch Transform run over on demand and reserved prices. A Fargate GPU Support probably should be cheaper. 

But I agree that for realtime inference, Sagemaker Serverless Inference is a good use case.

**Comment by uumami** [comment]
We want to run ML training of different models, sometimes a simple instance will suffice others we need GPUs and memory.

**Comment by appuCES** [comment]
+1, I'm using Fargate to process videos, I think having an accelerated hardware would help me cutting down some processing time.

**Comment by tltran-legion** [comment]
We are building a Cloud Engine, most of our workload are composed of video rendering using Nvidia drivers.  Building an ECS cluster with EC2 instances can be done, but not simple.  Having an ECS running on Fargate which support GPU will be much more efficient for  operational purpose.

**Comment by ngander-amfam** [comment]
Maybe with this announcement, https://aws.amazon.com/about-aws/whats-new/2022/03/bottlerocket-support-gpu-ec2-instance-types-powered-by-nvidia/, this can finally be realized.

**Comment by omieomye** [comment]
Quick update: we've begun thinking about how to ship this, and are actively designing. I'll leave this in the current state until we have better line-of-sight internally, and then provide another update.

**Comment by bigclap** [comment]
@omieomye Hi) Any news?

**Comment by omieomye** [comment]
@bigclap I don’t have a definitive when we’ll release this, but we’re actively implementing. A decision we have is what GPU type to support at launch. If the community has an opinion on needs GPU type(s), love to hear it.

**Comment by orgoro** [comment]
- g5
- g4dn
- p3

**Comment by tadam98** [comment]
Strongly support !

**Comment by ssathasivanACR** [comment]
GPU  on Fargate could be a game changer... Strongly support

**Comment by tltran-legion** [comment]
Thank you for the update @omieomye, we are actually using the g4dn actively on ca-central-1 and us-east-1.

**Comment by tbobm** [comment]
Thanks for the update @omieomye ! :bowing_man: 

I'm mostly using **g4dn** instances :)

**Comment by nexarChile** [comment]
THanks @omieomye for the good news!

At the moment we use **g4dn.xlarge** and **t2.small** instances on us-east-1, for processing video with GPU containers.

It will be very useful if we can use GPU containers with Fargate!
Please let us know when we can start using it. :D

**Comment by javabudd** [comment]
g5xlarge here!

**Comment by samevision** [comment]
I am really interested in this as well!

Could you give us any information when it will be available, @omieomye?

**Comment by renanmb** [comment]
On some of NVIDIA documentation there is stuff refering to this so. Its coming really soon

**Comment by aashitvyas** [comment]
We are currently utilizing all the G instances on our stack and interested in this feature as well.

**Comment by TejaswiniiB** [comment]
Hi @omieomye ,
Any update reg. when GPU support on Fargate will be available?

**Comment by RaiaN** [comment]
Support of Windows instances with GPUs is very much needed for Fargate.

**Comment by longerHost** [comment]
The GPU support will make fargate a great game-changer product. Thanks, @omieomye for the update, can't wait to see its release.

**Comment by deRooij** [comment]
This would help us out a lot, any status updates you are willing to share?
Our team uses mostly g4's, p3's and g5's.

**Comment by haugstve** [comment]
I was considering Fargate for hosting my model. Now it looks like I will install docker on EC2 and run it in a container that way. I hope this will be available soon.

**Comment by dvasilen** [comment]
@omieomye

> If the community has an opinion on needs GPU type(s), love to hear it.

We'd need A100 or at least V100.

**Comment by ltrojan** [comment]
+1

Any update on this issue? @omieomye ?

**Comment by omieomye** [comment]
Thanks for the feedback on GPU types! As is customary, I can't update on a specific month or date except to say that this is a top 5 capability enhancement for us and we're actively working on it.

**Comment by granthamtaylor** [comment]
Just wanted to add my +1 here.

GPU support for Fargate would be an absolute game changer for serverless ML build pipelines.

**Comment by tuanqle** [comment]
+1

**Comment by bieblebrox** [comment]
+1 👍

**Comment by johnatr** [comment]
Supporting GPU instances in Fargate would absolutely catapult not only Fargate itself but the whole serverless movement. Just scaffold API Gateway + Lambda Functions + Fargate and you are going to end up having lot of people dump Kubernetes.

**Comment by ktg0210** [comment]
+1

**Comment by hoduulmu** [comment]
+1

**Comment by mrifni** [comment]
@omieomye in the roadmap it says "49 We're Working On It", do you know when would this be available ?

**Comment by anna-geller** [comment]
+1 Adding GPU support to Fargate would make serverless finally applicable to data science workloads. This is really needed and would be so extremely appreciated!

**Comment by JunjieTang-D1** [comment]
+1 Important feature for Autonomous Driving Data Framework (ADDF) https://github.com/awslabs/autonomous-driving-data-framework

**Comment by ebr** [comment]
@heukirne where have you heard this? If on the AWS ML/AI innovation day twitch stream, that was only someone speculating in chat. There does not seem to be any concrete roadmap for this as far as I can tell. Unfortunately.

**Comment by Gunnar-Stunnar** [comment]
Is this what everyone is looking for?
https://aws.amazon.com/blogs/containers/running-gpu-based-container-applications-with-amazon-ecs-anywhere/

**Comment by appuCES** [comment]
> Is this what everyone is looking for?
> https://aws.amazon.com/blogs/containers/running-gpu-based-container-applications-with-amazon-ecs-anywhere/

I don’t think I’m looking for this.

**Comment by FelixRelli** [comment]
Using [AWS Batch](https://aws.amazon.com/de/batch/) to run on demand workloads using GPU instances might be a temporary workaround until fargate support is there.

**Comment by hervenivon** [comment]
> Using [AWS Batch](https://aws.amazon.com/de/batch/) to run on demand workloads using GPU instances might be a temporary workaround until fargate support is there.

Be careful, latency can be high with AWS Batch

**Comment by ofirnk** [comment]
Is there some quiet thread to subscribe to?
So that when there's an official update we'll get

**Comment by nselvakkumaran** [comment]
> Using [AWS Batch](https://aws.amazon.com/de/batch/) to run on demand workloads using GPU instances might be a temporary workaround until fargate support is there.

AWS Batch recommends using EC2 when a GPU is required. (i.e. no different from ECS task plus EC2 machines for capacity). 
https://docs.aws.amazon.com/batch/latest/userguide/fargate.html

Fargate needs to natively support GPU for realizing the goal of Serverless ML inferencing (most deep learning tasks).

**Comment by achaiah** [comment]
Really? No GPU support in 2023?

**Comment by genifycom** [comment]
Crazy isn't it. We need the GPU compute power in a serverless form. We needed it several years ago!

**Comment by RogerSanders** [comment]
Adding my +1 here to say this is something I also want to be able to do. When developing rich cloud-based applications, we have a diverse mix of plain CPU and GPU-accelerated compute tasks to execute, which are spun up on-demand by user actions. From our perspective, we just want to be able to launch all these through Fargate, not shunt the tasks that happen to be GPU accelerated through an entirely different pipeline.

**Comment by miolini** [comment]
I bet it will be released after the AGI was created event. :-)

**Comment by craigjbass** [comment]
Public sector could really benefit from GPU fargate

**Comment by ichenjia** [comment]
Any update on this?

**Comment by oscarnevarezleal** [comment]
Omg!! every ~1 year or so I fall into the same rabbit hole because I completely forgot about this limitation until I find this entry and then all come back to me. Still surprising that this request is 4 years old already.

**Comment by przemekblasiak** [comment]
+1

**Comment by shw1n** [comment]
Same, want to run AI inference on existing models (not even train)

Looks like we'll have to explore ECS

**Comment by karkir0003** [comment]
Any update on this? This would be super convenient!

**Comment by karkir0003** [comment]
@PettitWesley, @abby-fuller can you please comment on this issue ASAP!

**Comment by SuitoYoshiki** [comment]
+1

**Comment by ekaj2** [comment]
The importance of this is higher now than ever before :)

**Comment by BradVidler** [comment]
+1 very interested in this. Having to use EC2 instances that take 5-10 minutes to start is really unfortunate for our use case. There are competitors that can run instantly but our entire ecosystem is AWS based.

**Comment by sd031** [comment]
Eagerly waiting for this!

**Comment by buechera** [comment]
+1. Another team (AWS Panorama) with an interest in fargate GPU instances for our build/test pipelines.

**Comment by patradinesh** [comment]
+1

**Comment by jwstegemann** [comment]
+1

**Comment by DannySmrt** [comment]
+1

**Comment by bianchiidentity** [comment]
The current best practice is to launch GPU instances by executing ECS from lambda, right?

**Comment by killmepete** [comment]
+1

**Comment by JacekKosciesza** [comment]
+1
Use case: running panoramic image stitching software

**Comment by whillas-yabble** [comment]
just google "serverless GPU". There are lots of cheap third-party providers now that make it very easy to spin up instances and pay for what you use. Generally AWS is way behind when it comes to AI.

**Comment by simjak** [comment]
> The current best practice is to launch GPU instances by executing ECS from lambda, right?
> 
> 

Can you give a reference for this?

**Comment by raheem-imcapsule** [comment]
Any update on this

**Comment by nonedone** [comment]
Adding my nudge here on this historic ticket.

**Comment by backnight** [comment]
+1

**Comment by dscain** [comment]
For simple Generative AI workloads, fargate+gpu would be really valuable, even more now with all with so much people trying gen AI applications.

**Comment by oscarnevarezleal** [comment]
The reality is there's not enough GPU power available for on demand scenarios. Even more, the available supply goes directly to big players.

I just opened some AWS accounts for new clients and they start with a zero EC2 GPU quota, have to increasing by submitting an increment request and even after that they don't give you all the availability you ask for rather they told you to wait and see if you really needed more.

Looking at the trends over the years GPU will need to become either absurdly cheap or widely available (same thing must of the times ) before we can go on demand mode for later to be IaC available.

**Comment by kunal14053** [comment]
Do we have any updates on this?

**Comment by genifycom** [comment]
Sadly Amazon are letting us down on the AI/ML front by not giving us the flexibility we need to advance. As a result we are falling behind where we should be at this point.

The model seems to be that cloud providers know best and will force us down their path.

We need GPU access in multiple scenarios.

**Comment by FelixRelli** [comment]
> Custom Model Import was announced for Bedrock the other day. Not sure of all your use cases, but might be an option for a managed AI model hosting with pay-per-token pricing. https://aws.amazon.com/about-aws/whats-new/2024/04/custom-model-import-amazon-bedrock/

"Custom models can only be accessed using Provisioned Throughput." => So no pay-per token.
https://aws.amazon.com/bedrock/pricing/

**Comment by kmulka-bloomberg** [comment]
> "Custom models can only be accessed using Provisioned Throughput." => So no pay-per token. https://aws.amazon.com/bedrock/pricing/

I'm working on getting clarification from AWS, but this blog post says that custom model import uses the On-Demand mode.
https://aws.amazon.com/blogs/aws/import-custom-models-in-amazon-bedrock-preview/

**Comment by tamreddy-dot** [comment]
.

**Comment by ssignal** [comment]
@johnwheeler
I think the error message and the below link will help you.
"Resource handler returned message: "Invalid request provided: Create TaskDefinition: Tasks using the Fargate launch type do not support GPU resource requirements."

https://nocd.hashnode.dev/registering-gpu-instance-w-aws-elastic-container-service-ecs

**Comment by kendrexs** [comment]
This is WIP for ECS Fargate

**Comment by kendrexs** [comment]
updated commend to clarify the feature for GPUs and other advanced capacity

**Comment by omers** [comment]
+1
👀

**Comment by marksverdhei** [comment]
Someone tell the executives how much money they would've made if this had been implemented in 2020. Still missing this!

**Comment by yogendratamang48** [comment]
We are waiting for this. When is this coming?

**Comment by seghcder** [comment]
We've been waiting 18,168x longer than they spent Waiting for Godot...

**Comment by Abdelwahab313** [comment]
Hi team,

First of all, thanks for the work on this — I'm really looking forward to the upcoming support for GPU on Fargate.

However, I’d also like to highlight that for our use case (LLM hosting and background processing), having proper support for graceful termination is equally critical. At the moment, the hard limit of stopTimeout to 2 minutes poses a real challenge when autoscaling Fargate tasks, often leading to incomplete processing or data loss on shutdown.

We’re currently relying on EC2 because it gives us the flexibility we need — both GPU support and customizable termination handling. Fargate would be a perfect fit for us if both of these capabilities were available.

Please consider this a +1 for supporting longer/more configurable stopTimeout values as part of the ongoing enhancements. Here's the related issue: #1020.

Thanks again for your efforts!

**Comment by NonpareilNic** [comment]
As a long time corporate user of ECS Fargate and now building MLOps, etc. I'd absolutely love GPUs via Fargate and any updates you can give on it.

**Comment by jorlando-elucid** [comment]
> Someone tell the executives how much money they would've made if this had been implemented in 2020. Still missing this!

Unfortunately, that is not how market forces work. That would require people to have not run GPU based workloads they other wise would have. In reality, people still ran these GPU workloads, just on over provisioned GPU based EC2 nodes. 

Until there is a real alternative service providing ephemeral GPU access to containers that would have other wise run on AWS, AWS isn't missing out on any money. In fact, they may make more with the over provisioned EC2 nodes.

**Comment by mikestaub** [comment]
I have been running all my GPU workloads on beam.cloud, I would have definitely used Fargate if it was available as all the rest of my infra is on AWS.

**Comment by geovanisouza92** [comment]
> Until there is a real alternative service providing ephemeral GPU access to containers that would have other wise run on AWS, AWS isn't missing out on any money.

We've been using https://www.runpod.io/ because it's cheaper.

**Comment by kendrexs** [comment]
General Reminder/Update on my comment above: this feature will not look quite like Fargate GPUs as such, rather it is a new capacity type; https://github.com/aws/containers-roadmap/issues/2452

**Comment by aniketw** [comment]
We are requesting GPU support for both ECS Fargate and EKS Fargate services to support our use case involving 3D map rendering. This process requires GPU and OpenGL capabilities to generate videos or images of rendered maps with custom graphics layers.
Fargate is our preferred compute option due to its serverless nature, which helps us avoid managing the underlying infrastructure. This is especially important for us given the compliance and security requirements we must adhere to.
Currently, we are using EC2-backed ECS instances for development purposes. However, clarity on whether GPU support is planned for Fargate would greatly help us in planning our architecture and future deployments.

**Comment by genifycom** [comment]
I support others comments. Could really use an intermittent GPU serverless solution.

GPU on demand!

**Comment by jsm** [comment]
FYI They just announced this: https://aws.amazon.com/blogs/aws/announcing-amazon-ecs-managed-instances-for-containerized-applications/

**Comment by dfuentes77** [comment]
Yeah. The announcement was not what I was hoping for. I was looking forward to the solution leaning more on the Fargate side and not on the EC2 side.

**Comment by azN2** [comment]
> Yeah. The announcement was not what I was hoping for. I was looking forward to the solution leaning more on the Fargate side and not on the EC2 side.

you realize fargate was just "managed" ec2 instances right?

**Comment by jimjam-slam** [comment]
> you realize fargate was just "managed" ec2 instances right?

In our case, we're using SST v3 to manage resources, and would be very interested in using that to deploy GPU workloads (as it uses ECS with Fargate, not ECS with EC2). Of course we can deploy with ECS + EC2 ourselves (and will probably look at that when it comes time to do that work), but it'd certainly be nice if we could manage it all in SST through Fargate!

Nevertheless, since this is tracked separately to the ECS Managed Instances and is also marked as 'coming soon', I'm crossing my fingers 🤞🏻

**Comment by AbhishekNautiyal** [comment]
This feature has been delivered with the Amazon ECS Managed Instances launch as covered in https://github.com/aws/containers-roadmap/issues/2452. You can specify your desired instance requirements, including GPUs, for fully managed compute with ECS Managed Instances. Refer to our [documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/managed-instances-instance-selection-best-practices.html) to learn more.

**Comment by rhamnett** [comment]
Cheers for the quick turnaround on that.

**Comment by gautaz** [comment]
Hello @AbhishekNautiyal, thanks for the feedback.
Could you also tell us how this will be reflected in CloudFormation recipes?
For instance, how does one configure a Batch [ComputeEnvironment](https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-resource-batch-computeenvironment.html) with CFN to benefit from this new feature?

**Comment by jasonheffnerpsu** [comment]
@AbhishekNautiyal I have the same question as @gautaz . We currently use AWS Batch to submit Fargate jobs because it simplifies orchestration and lets us control timeouts and retries without managing any EC2 capacity. Do you know if ECS Managed Instances will also be supported within AWS Batch Compute Environments, or reflected in CloudFormation/CDK definitions? 

I had a similar issue when ECS Default Logging Behavior was changed and there is no separate setting for blocking/non-blocking in AWS Batch definitions. I set this globally and submitted a feature request for that issue in particular.

**Comment by KinderGuy21** [comment]
Does anyone have any updates on this? How soon is Coming Soon? This could help create more of a serverless for GPU related actions that need fast responses and not a lot of traffic, which is perfect for my use cases.

**Comment by pcgeek86** [comment]
Another AWS re:Invent has come and gone, and I don't believe GPUs are supported yet .... would be great if NVIDIA GPUs could be supported on AWS Fargate, similar to [Google Cloud Run](https://docs.cloud.google.com/run/docs/configuring/services/gpu).

Many of our customers run our [self-hosted Deepgram](https://developers.deepgram.com/docs/aws-k8s) voice AI containers on Amazon EKS clusters. We occasionally come across customers who specifically want to run Deepgram on AWS Fargate, as it requires less management overhead for their engineering teams.

Any news on this?

**Comment by michal-kosinski** [comment]
Hi,

Is there any way to get GPU metrics when using Amazon ECS Managed Instances? I haven't been able to find this information in the documentation[1].

[1] https://docs.aws.amazon.com/AmazonECS/latest/developerguide/additional-metrics-managed-instances.html

**Comment by AbhishekNautiyal** [comment]
> [@AbhishekNautiyal](https://github.com/AbhishekNautiyal) I have the same question as [@gautaz](https://github.com/gautaz) . We currently use AWS Batch to submit Fargate jobs because it simplifies orchestration and lets us control timeouts and retries without managing any EC2 capacity. Do you know if ECS Managed Instances will also be supported within AWS Batch Compute Environments, or reflected in CloudFormation/CDK definitions?
> 
> I had a similar issue when ECS Default Logging Behavior was changed and there is no separate setting for blocking/non-blocking in AWS Batch definitions. I set this globally and submitted a feature request for that issue in particular.

Hi @jasonheffnerpsu 

Thanks for sharing details about your use case.  Could you please open a feature request for Batch support for Managed Instances? We're discussing the prioritization of this capability with the AWS Batch team, and a separate thread will help us share updates more effectively.

**Comment by AbhishekNautiyal** [comment]
> Hi,
> 
> Is there any way to get GPU metrics when using Amazon ECS Managed Instances? I haven't been able to find this information in the documentation[1].
> 
> [1] https://docs.aws.amazon.com/AmazonECS/latest/developerguide/additional-metrics-managed-instances.html

@michal-kosinski 
We have plans for adding support for additional metrics including GPU monitoring for ECS Managed Instances.  Could you please open a separate feature request, so that we can track and share updates for this specific capability?

**Comment by AbhishekNautiyal** [comment]
We have delivered support for GPUs on fully-managed compute with the ECS Managed Instances launch (https://github.com/aws/containers-roadmap/issues/2452). ECS Managed Instances is a fully managed compute option that delivers the same operational and security benefits as Fargate while providing access to the full range of EC2 capabilities.

We are aware that this feature request was originally opened for Fargate. While Managed Instances addresses this need, some customers may still want access to GPUs on Fargate. I am closing this issue now so that we can focus and gather feedback for prioritization of Fargate-specific features. Please open a new issue if you require GPU/accelerator support on Fargate, including your use case(s), or if you have any other requests.  As always, we appreciate your engagement and feedback on this thread.

**Comment by michal-kosinski** [comment]
> [@michal-kosinski](https://github.com/michal-kosinski) We have plans for adding support for additional metrics including GPU monitoring for ECS Managed Instances. Could you please open a separate feature request, so that we can track and share updates for this specific capability?

@AbhishekNautiyal
Good to know that this feature is planned! I've created a feature request to track this capability: https://github.com/aws/containers-roadmap/issues/2734

**Comment by gautaz** [comment]
Hello, I've just created #2735 to describe our current use case regarding deploying GPU intensive Batch jobs with CloudFormation.

Feel free to comment and upvote if you share the same use case.
